packages/flutter_tools/lib/src/runner/flutter_command.dart

```diff
@@ -635,10 +635,22 @@ mixin DeviceBasedDevelopmentArtifacts on FlutterCommand {
           artifacts.add(DevelopmentArtifact.iOS);
           break;
         case TargetPlatform.darwin_x64:
-        case TargetPlatform.fuchsia:
-        case TargetPlatform.tester:
+          if (!FlutterVersion.instance.isStable) {
+            artifacts.add(DevelopmentArtifact.macOS);
+          }
+          break;
         case TargetPlatform.windows_x64:
+          if (!FlutterVersion.instance.isStable) {
+            artifacts.add(DevelopmentArtifact.windows);
+          }
+          break;
         case TargetPlatform.linux_x64:
+          if (!FlutterVersion.instance.isStable) {
+            artifacts.add(DevelopmentArtifact.linux);
+          }
+          break;
+        case TargetPlatform.fuchsia:
+        case TargetPlatform.tester:
           // No artifacts currently supported.
           break;
       }
```

Comment:
Given that the entire switch statement is identical between the two methods, how about extracting it to a helper method that takes the target platform as a param?